sdks/go/pkg/beam/runners/prism/internal/engine/strategy.go

```diff
@@ -23,16 +23,351 @@ import (
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+	// Track fired states.
+	if ts.extra == nil {
+		ts.extra = map[Trigger]bool{}
+	}
+	isFinished := ts.extra.(map[Trigger]bool)
+	for _, t := range t.SubTriggers {
+		// Don't re-evaluate.
+		if isFinished[t] {
+			continue
+		}
+		if t.isReady(input, state) {
+			isFinished[t] = true
+		}
+	}
+	ready := len(isFinished) == len(t.SubTriggers)
+	if ready {
+		ts.finished = true
+	}
+	state.setTriggerState(t, ts)
+	return ready
+}
+
+func (t *TriggerAfterAll) reset(state *StateData) {
+	for _, sub := range t.SubTriggers {
+		sub.reset(state)
+	}
+	delete(state.Trigger, t)
+}
+
+// TriggerAfterAny is ready the first time any of the subTriggers are ready.
+// Logically, an "OR" trigger.
+type TriggerAfterAny struct {
+	SubTriggers []Trigger
+}
+
+func (t *TriggerAfterAny) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+	anyReady := false
+	for _, t := range t.SubTriggers {
+		anyReady = t.isReady(input, state)
+		if anyReady {
+			break
+		}
+	}
+	if anyReady {
+		ts.finished = true
+		for _, sub := range t.SubTriggers {
+			// clear any subtrigger state, as we're fireing anyway.
+			sub.reset(state)
+		}
+	}
+	state.setTriggerState(t, ts)
+	return anyReady
+}
+
+func (t *TriggerAfterAny) reset(state *StateData) {
+	for _, sub := range t.SubTriggers {
+		sub.reset(state)
+	}
+	delete(state.Trigger, t)
+}
+
+// TriggerAfterEach processes each trigger before executing the next.
+// Starting with the first subtrigger, ready when the _current_ subtrigger
+// is ready. After output, advances the current trigger by one.
+type TriggerAfterEach struct {
+	SubTriggers []Trigger
+}
+
+func (t *TriggerAfterEach) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+	// Use extra for the current trigger index.
+	if ts.extra == nil {
+		ts.extra = int(0)
+	}
+	current := ts.extra.(int)
+	ready := false
+	if t.SubTriggers[current].isReady(input, state) {
+		current++
+		ready = true
+	}
+	if current >= len(t.SubTriggers) {
+		ts.finished = true
+	}
+	ts.extra = current
+	state.setTriggerState(t, ts)
+	return ready
+}
+
+func (t *TriggerAfterEach) reset(state *StateData) {
+	for _, sub := range t.SubTriggers {
+		sub.reset(state)
+	}
+	delete(state.Trigger, t)
+}
+
+// TriggerElementCount triggers when there have been at least the required number
+// of elements have arrived.
+type TriggerElementCount struct {
+	ElementCount int
+}
+
+func (t *TriggerElementCount) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+
+	if ts.extra == nil {
+		ts.extra = int(0)
+	}
+	count := ts.extra.(int) + input.newElementCount
+	ts.extra = count
+	ready := count >= t.ElementCount
+	if ready {
+		ts.finished = true
+	}
+	state.setTriggerState(t, ts)
```

Comment:
This is benign but also not necessary. Just default it to 0 in `onElement`.

Reply:
It is necessary since triggerState is passed by value and not by pointer at present. It gains more reasonable zero value semantics, since we don't have to do both a single value initialization, *and* any extra value initialization. It also makes the state handling explicit.